homeassistant/components/airscape/fan.py

```diff
@@ -0,0 +1,173 @@
+"""Platform for Airscape fan integration."""
+import logging
+import airscape
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+
+# Import the device class from the component that you want to support
+from homeassistant.components.fan import (
+    SUPPORT_SET_SPEED,
+    PLATFORM_SCHEMA,
+    FanEntity,
+    DOMAIN,
+)
+from homeassistant.const import CONF_HOST, CONF_NAME, CONF_TIMEOUT, CONF_MINIMUM
+
+DEFAULT_TIMEOUT = 5
+DEFAULT_MINIMUM = 1
+
+_LOGGER = logging.getLogger(__name__)
+
+# Validation of the user's configuration
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_NAME): cv.string,
+        vol.Optional(CONF_TIMEOUT, default=DEFAULT_TIMEOUT): cv.positive_int,
+        vol.Optional(CONF_MINIMUM, default=DEFAULT_MINIMUM): cv.positive_int,
+    }
+)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Airscape Fan platform."""
+    # Setup connection to the fan
+    host = config.get(CONF_HOST)
+    name = config.get(CONF_NAME)
+    minimum = config.get(CONF_MINIMUM)
+    timeout = config.get(CONF_TIMEOUT)
+
+    try:
+        device = airscape.Fan(host, timeout)
+    except (airscape.exceptions.Timeout, airscape.exceptions.ConnectionError):
+        _LOGGER.error(
+            "Cannot connect to %s.  " "Device did not respond to API request", host
+        )
+    else:
+        # Add devices
+        add_entities([AirscapeWHF(device, name, minimum)], True)
+
+    def service_speed_up(call):
+        """Handle speed_up service call."""
+        entity_id = call.data.get("entity_id")
+        _LOGGER.debug("Calling speed_up for %s", entity_id)
+
+        entity = hass.data[DOMAIN].get_entity(entity_id)
```

Comment:
Please store the entities specifically in hass.data and access them from that. We can't rely on the entity component helper storing them. That's an implementation detail of the helper.

Reply:
Then how do I associate the `entity_id` of the call with the object I want to manipulate?  It isn't guaranteed that there will only be one instance of this platform, so accessing `hass.data[DOMAIN]["device"]` for example may not return the correct entity object.  And since `entity_id` is not available until after the `setup_platform` has completed, I don't have a way to reference it during `setup_platform` to add it to `hass.data` structure like `hass.data[DOMAIN][entity_id]`

Reply:
We can store the entity in hass.data in `async_added_to_hass` entity coroutine method. The entity_id of the entity is available at that point.

An alternative is to not interact directly with the entity but use our dispatch helper to signal the correct entity method from the service handler. This is more future proof since we can move handlers out of the platform if necessary if we want to expand to a component with platforms in the future.

Reply:
I used your suggestion of using the dispatch helper.  I struggled a bit with how to get access to the entity_id in `async_added_to_hass`.  Then I discovered that `entity_id` gets added to the object and is accessible at `self.entity_id`.  That isn't documented anywhere.  That would be helpful to have written down somewhere.  Maybe a doc on what gets added to an object once it gets added with `setup_platform` would be useful.  There is somewhere in the dev docs that mentions that the hass object gets added, but no notes on anything else.

Also, there are no docs for `homeassistant.helpers.dispatcher` anywhere that I could find on https://dev-docs.home-assistant.io/en/master/api/